### PR TITLE
Life Line Charts / Place Coordinates

### PR DESCRIPTION
**Pull request by CWSchulze**
1. There is a new View "life_line_chart" (see https://github.com/CWSchulze/life_line_chart / https://pypi.org/project/life-line-chart) - this is still being developed
    - A ancestor chart which shows the actual life span, birth and death dates, and the relation between the individuals
    - The ancestor collapse is addressed
2. A Gramplet/Geography View to simplify searching for addresses and geographic locations
    - You can click on the map, and add new places in the popup menu. City names etc. are automatically detected. 
    - You can click on a place in the Places View and search for a geographic location using the gramplet
3. Godfather as event role
4. Short guideline to setup msys2/mingw under windows to develop gramps

Some feedback would be nice.

### CODE REVIEW

**Comment by sam-m888** [comment]
Thank you for the new chart.

Note that Travis-Ci failed due to trailing white spaces see [here](https://travis-ci.org/github/gramps-project/gramps/builds/680440246#L2895)

Your chart example reminds me of the TimeNet Chart mentioned here on the Gramps wiki back in 2010.
*  [GEPS 030: New Visualization Techniques - Gramps](https://www.gramps-project.org/wiki/index.php?title=GEPS_030:_New_Visualization_Techniques#TimeNets)

Please consider moving this to the Gramps addons as it has a better chance of begin tested and if any issues arise  with the addon you can rapidly push out fixes instead of waiting for a Gramps point release.

**Comment by SNoiraud** [comment]
I saw you needs the geopy interface which add 4 or 5 dependencies to gramps.
Why you didn't use geocodeglib which is already used by gramps for the place selection ?

**Comment by CWSchulze** [comment]
What you mean by Gramps addons? Do you mean https://github.com/gramps-project/addons-source?
How is it possible to access these addons? Is it possible to add modules to some requirements.txt?
Place Coordinate and Life Line Charts can be moved to addons.

I didn't know where to place the guideline to setup msys2 on windows.

Is the TimeNet chart still available? It seems to be similar but the example looks very clean and simple. Life Line Chart also addresses complex cross dependencies. E.g.:
grafik


Yes, my understanding of pretty formatting seems to be different from pep 8. I will try to auto format this.

**Comment by CWSchulze** [comment]
> I saw you needs the geopy interface which add 4 or 5 dependencies to gramps.
> Why you didn't use geocodeglib which is already used by gramps for the place selection ?

I will look into this. I wrote this addon 6 years ago...
Geopy worked pretty well and found also also ugly formatted addresses. It provides search in both directions. Maybe geocodeglib also does, I have to check that.

**Comment by wroldwiedbwe** [comment]
Looks good thanks for creating them 👍 

Attempted converting both to addons  ([CWSchulze_Life Line Charts  and Place Coordinates.zip](https://github.com/gramps-project/gramps/files/4545558/CWSchulze_Life.Line.Charts.and.Place.Coordinates.zip)
)for myself to test and managed to only install geopy without issue the other package life_line_chart gave errors so no testing from me (error messages further down). 

2020-04-28 19:58:41.517: WARNING: _manager.py: line 308: Plugin error (from 'PlaceCoordinateGeoView'): No module named 'geopy'

pip3 install geopy

Pics for what I could get working :

Place Coordinates Gramplets
Place Coordinates Gramplets

GeoPlace Coordinate Gramplet view
GeoPlace Coordinate Gramplet view

For the "Life Line Charts" had issues getting your requirement installed on MSYS2

> 2020-04-28 19:58:37.691: WARNING: _manager.py: line 308: Plugin error (from 'lifelinechartview'): No module named 'life_line_chart'


pip3 install life_line_chart

> $ pip3 install life_line_chart
> Collecting life_line_chart
>   Using cached life_line_chart-1.2.10-py3-none-any.whl (47 kB)
> Collecting svgpathtools
>   Downloading svgpathtools-1.3.3-py2.py3-none-any.whl (50 kB)
>      |################################| 50 kB 1.6 MB/s
> Collecting numpy
>   Using cached numpy-1.18.3.zip (5.4 MB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>     Preparing wheel metadata ... error
>     ERROR: Command errored out with exit status 1:
>      command: c:/msys64/mingw64/bin/python.exe C:/msys64/mingw64/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py prepare_metadata_for_build_wheel C:/Users/wroldwiedbwejr/AppData/Local/Temp/tmpl__rrese
>          cwd: C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy
>     Complete output (86 lines):
>     Processing numpy/random/_bounded_integers.pxd.in
>     Processing numpy/random/mtrand.pyx
>     Processing numpy/random/_bit_generator.pyx
>     Processing numpy/random/_bounded_integers.pyx.in
>     Processing numpy/random/_common.pyx
>     Processing numpy/random/_generator.pyx
>     Processing numpy/random/_mt19937.pyx
>     Processing numpy/random/_pcg64.pyx
>     Processing numpy/random/_philox.pyx
>     Processing numpy/random/_sfc64.pyx
>     Cythonizing sources
>     blas_opt_info:
>     blas_mkl_info:
>     'gcc' is not recognized as an internal or external command,
>     operable program or batch file.
>     Running from numpy source directory.
>     setup.py:461: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates
>       run_build = parse_setuppy_commands()
>     Traceback (most recent call last):
>       File "C:/msys64/mingw64/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 257, in <module>
>         main()
>       File "C:/msys64/mingw64/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 240, in main
>         json_out['return_val'] = hook(**hook_input['kwargs'])
>       File "C:/msys64/mingw64/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 110, in prepare_metadata_for_build_wheel
>         return hook(metadata_directory, config_settings)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-build-env-wu8c11xl/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 158, in prepare_metadata_for_build_wheel
>         self.run_setup()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-build-env-wu8c11xl/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 249, in run_setup
>         super(_BuildMetaLegacyBackend,
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-build-env-wu8c11xl/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 143, in run_setup
>         exec(compile(code, __file__, 'exec'), locals())
>       File "setup.py", line 488, in <module>
>         setup_package()
>       File "setup.py", line 480, in setup_package
>         setup(**metadata)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/core.py", line 137, in setup
>         config = configuration()
>       File "setup.py", line 165, in configuration
>         config.add_subpackage('numpy')
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/misc_util.py", line 1033, in add_subpackage
>         config_list = self.get_subpackage(subpackage_name, subpackage_path,
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/misc_util.py", line 999, in get_subpackage
>         config = self._get_configuration_from_setup_py(
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/misc_util.py", line 941, in _get_configuration_from_setup_py
>         config = setup_module.configuration(*args)
>       File "numpy/setup.py", line 10, in configuration
>         config.add_subpackage('core')
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/misc_util.py", line 1033, in add_subpackage
>         config_list = self.get_subpackage(subpackage_name, subpackage_path,
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/misc_util.py", line 999, in get_subpackage
>         config = self._get_configuration_from_setup_py(
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/misc_util.py", line 941, in _get_configuration_from_setup_py
>         config = setup_module.configuration(*args)
>       File "numpy/core/setup.py", line 759, in configuration
>         blas_info = get_info('blas_opt', 0)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 471, in get_info
>         return cl().get_info(notfound_action)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 728, in get_info
>         self.calc_info()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 1896, in calc_info
>         if self._calc_info(blas):
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 1875, in _calc_info
>         return getattr(self, '_calc_info_{}'.format(name))()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 1812, in _calc_info_mkl
>         info = get_info('blas_mkl')
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 471, in get_info
>         return cl().get_info(notfound_action)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 728, in get_info
>         self.calc_info()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 1179, in calc_info
>         info = self.check_libs2(lib_dirs, mkl_libs)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 884, in check_libs2
>         exts = self.library_extensions()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 843, in library_extensions
>         c = customized_ccompiler()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/system_info.py", line 185, in customized_ccompiler
>         global_compiler = _customized_ccompiler()
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/__init__.py", line 51, in customized_ccompiler
>         c = ccompiler.new_compiler(plat=plat, compiler=compiler, verbose=verbose)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/ccompiler.py", line 771, in new_compiler
>         compiler = klass(None, dry_run, force)
>       File "C:/Users/wroldwiedbwejr/AppData/Local/Temp/pip-install-vwhh90om/numpy/numpy/distutils/mingw32ccompiler.py", line 68, in __init__
>         distutils.cygwinccompiler.CygwinCCompiler.__init__ (self, verbose,
>       File "C:/msys64/mingw64/lib/python3.8/distutils/cygwinccompiler.py", line 127, in __init__
>         if self.ld_version >= "2.10.90":
>     TypeError: '>=' not supported between instances of 'NoneType' and 'str'
>     ----------------------------------------
> ERROR: Command errored out with exit status 1: c:/msys64/mingw64/bin/python.exe C:/msys64/mingw64/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py prepare_metadata_for_build_wheel C:/Users/wroldwiedbwejr/AppData/Local/Temp/tmpl__rrese Check the logs for full command output.
> 

Did not test:
* Godfather as event role
or read
* Short guideline to setup msys2/mingw under windows to develop gramps

Will keep test whats working for me and report back with issues

Yeah to much happening on the PR needs to be split etc..

**Comment by CWSchulze** [comment]
@wroldwiedbwe : I also failed to compile some modules on msys2/mingw. However, you can use the binary packages with pacman:
```pacman -S mingw-w64-x86_64-python-dateutil mingw-w64-x86_64-python-numpy mingw-w64-x86_64-python-pillow```
-> https://packages.msys2.org

In the PlaceCoordinate gramplet you can press OK to search for a geographic location for the place you have selected:
grafik
Im your case searching for the geographic location of "USA" is not really useful.
After something has been found, the geographic location can be applied to the place by the bottom right button.

To add new places you can use the geography view:
grafik
Name and postal code are autodetected:
grafik

**Comment by prculley** [comment]
It would be nice if this didn't need numpy as a prerequisite.  If we have to add that to the Windows AIO, it would just about double the size of the installation files.  And unlike on Linux, the user cannot easily add prerequisites on his own later.

**Comment by CWSchulze** [comment]
I think numpy is required by the svgpathtools module. This is mainly required to place fotos of individuals on the life lines (to compute the intersection of a spline and a line). Well, and it is used to draw text along a spline with cairo - which is utterly slow anyhow, so I don't use it in gramps.

I think there are ways to get a small binary for numpy which was compiled with gcc instead of the cpu specific mkl binaries compiled with intel (?). Maybe it is easier to drop this dependency.

**Comment by prculley** [comment]
I guessed as much.  I wonder if it might be easier to just use a standardized curved (radius) connector and simple lines.  It would not look quite as nice, but should still work well.  Just thoughts; I really don't know anything about SVG...

**Comment by CWSchulze** [comment]
In SVG you define splines by setting anchor / control points, you can even link text to a path - no problem with that, its simple and nothing has to be calculated in advance, so numpy is not required for this.

But it is complicated to precalculate where the lines will be shown, calculate intersections, i.e. manually place additional items along the line. Since cairo does not support rendering text along a path, I had to find a workaround to replace this build-in SVG feature. This workaround for rendering in gramps calculates intersections with the splines to place single characters.

However, rendering the birth label along a path is optional, pretty slow in gramps and not that much of a benefit.
I don't know yet how use the optional individual-fotos-along-the-line-feature in gramps.

So basically numpy is not required.

**Comment by prculley** [comment]
I realize it is early days yet, however, you should look into getting the code to mostly comply with PEP8.  We have a standard which basically says each new file must have a pylint score of 90% or higher to be accepted into Gramps.  The Codacy/PR Quality review referenced at the end of this PR is basically a 'pylint like' report, although it doesn't indicate the %.  You have to use the actual pylint utility for that.

Some of the 'standard' ways we have traditionally done our Gramps files are not quite up to PEP8 standards (things like the order of imports, two newlines before classes etc.) so it is difficult to achieve 100%.

File: debugOnWindowsWithVSCode.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    - https://packages.msys2.org
+
+- use zsh! https://www.tfzx.net/article/2750894.html
+
+# Life Line Chart
+- pacman -S mingw-w64-x86_64-python-dateutil mingw-w64-x86_64-python-numpy mingw-w64-x86_64-python-pillow
+- pip install svgwrite svgpathtools
+
+# VSCode
+
```
Comment (prculley) [inline debugOnWindowsWithVSCode.md:19]: Out of idle curiosity, what does VSCode have to do with msys2/mingw or this addon?  I suggest some additional text describing what/why you need the items below.  Or maybe make a wiki article about your choice of IDE and how to debug with Gramps on Windows.

File: gramps/gen/lib/eventroletype.py [hunk 2]
```diff
@@ -60,6 +61,7 @@ class EventRoleType(GrampsType):
         (WITNESS, _("Witness"), "Witness"),
         (FAMILY, _("Role|Family"), "Family"),
         (INFORMANT, _("Informant"), "Informant"),
+        (GODFATHER, _("Godfather"), "Godfather"),
         ]
```
Comment (prculley) [inline gramps/gen/lib/eventroletype.py:65]: Is this change critical to using the addon?  Maybe a separate PR?

File: gramps/gen/lib/place.py [hunk 3]
```diff
@@ -576,10 +576,16 @@ def _merge_alt_names(self, acquisition):
...
+        # if there is a acquisition name,
+        #  and the length of the acquisition name is > 0,
+        #  and the names of the acquisition and the original place are not identical,
+        #  and the acquisition is not yet in the list of alternate names
+        # then append the acquisition to the list of alternate name
+        if acquisition.name and \
+                (len(acquisition.name.value)>0) and \
+                (acquisition.name.value != self.name.value) and \
+                (acquisition.name not in self.alt_names):
+            self.alt_names.append(acquisition.name)
```
Comment (prculley) [inline gramps/gen/lib/place.py:588]: Is this change critical to using the addon?

**Comment by prculley** [comment]
> What you mean by Gramps addons? Do you mean https://github.com/gramps-project/addons-source?
> How is it possible to access these addons? Is it possible to add modules to some requirements.txt?
> Place Coordinate and Life Line Charts can be moved to addons.
That is the correct repo for the Gramps addons.  As with Gramps, clone the addons-source repo to your Github account, create a branch based off of maintenance/gramps51 with your new addons, push to your Github account, and then create a PR for Gramps addons-source.  I suggest basing off of gramps51 because you could release it when done, rather than waiting until master is released as a new version of Gramps.
> 
> I didn't know where to place the guideline to setup msys2 on windows.
I don't think this is really a necessary part of the addon; it seems more like a wiki article similar to https://gramps-project.org/wiki/index.php/Gramps_for_Windows_with_MSYS2.

If you end up needing additional dependencies, you should check that they are present in your gpr.py code similar to how it is done in https://github.com/gramps-project/addons-source/blob/master/GraphView/graphview.gpr.py.  We can also add them to the list in the Prerequisites Checker Gramplet once the addon is available.

**Comment by CWSchulze** [comment]
1. I am working on splitting this pull request. Ongoing work.
2. The VSCode and setup guideline is point of discussion. I suggest to link the wiki-articles about setting up an environment on the github main readme.md.
I wasted a lot of time setting up some python enabling me to debug gramps 4.x several years ago, and again version 5.x more recently. However, there seems to be a guidline for msys2 on the website which i didn't know when working out that one. I was not aware that msys2 was the buzzword I was searching for - usually I work with conda and Qt.
Additionally there are some special issues with msys and vscode. But I wasn't able to solve all of them yet. Especially the tests are implemented different from what vscode is expecting. I wouldn't know where to place that information.
2. PEP8: In the Codacy report I get the worst ratings (F) for the .gpr.py files. But looking at other gramplets, that seems to be the way these files are supposed to look. I get errors for importing "gi.whatever" or my lifelinechart module. I will ignore those errors.
I will try to stick to your formatting, even though I personally find docstrings with argument description more important than the number of new lines or accessing protected members... But I guess thats an endless discussion.
3. Changes relating godfather eventrole and place are separate issues. The latter actually came with several changes for the gedcom import, but git and me failed to update them from version 4.2. The original setup for places prevented alternative names in general. I will need to find an example for that.
4. I cannot reproduce the travis ci problems. I don't see what i did wrong, and i cannot run the nosetests on my msys. There seem to be also language dependent issues in the tests:
grafik
However, I'm not sure if I really have to reproduce what is happening there - I'd rather not.

Thanks I will look into that example for the requirements, and create a pull request on addons-source.

**Comment by adamvazquez82** [comment]
PR Split out to:
* [PR#334 Place coordinate gramplet](https://github.com/gramps-project/addons-source/pull/334)
* [PR#335 Life line chart view](https://github.com/gramps-project/addons-source/pull/335)

**Comment by prculley** [comment]
Should we close this PR, or do you expect further work here and want to try to get it into the mainstream Gramps code?

**Comment by CWSchulze** [comment]
The place issue seems to be solved also with the current master. I'm working to merge the other changes regarding gedcom import.
I'll create a pull request for the godfather eventrole.

**Comment by uli22** [comment]
Hi!
I tried out the life line descendent chart and I like it.
One small point: when using with dark mode it would be nice to have white
letters and a lucent "wedding rings". see attached screenshot
Thanks
cheers
Uli
[image: image.png]

Am Sa., 2. Mai 2020 um 17:31 Uhr schrieb CWSchulze <notifications@github.com

**Comment by CWSchulze** [comment]
@uli22, thanks for the feedback.

Unfortunately I don't see your pictures, but I was able to install the themes addon and test the dark mode. I see the point. This can be fixed. Yet I don't know how, but I'm sure it can. ;)

The discussion moved to addon-source project and PR https://github.com/gramps-project/addons-source/pull/335
